url748e34f967_Diab1o_java_7z_archiver_tgz-pJ8-com_swemel_sevenzip_TestJ8/src/com/swemel/sevenzip/compression/rangecoder/BitTreeEncoder.java

```diff
@@ -0,0 +1,85 @@
+package com.swemel.sevenzip.compression.rangecoder;
+
+import java.io.IOException;
+
+public class BitTreeEncoder {
+    private final short[] Models;
+    private final int NumBitLevels;
+
+    public BitTreeEncoder(int numBitLevels) {
+        NumBitLevels = numBitLevels;
+        Models = new short[1 << numBitLevels];
+    }
+
+    public void init() {
+        Decoder.initBitModels(Models);
+    }
+
+    public void encode(Encoder rangeEncoder, int symbol) throws IOException {
+        int m = 1;
+        for (int bitIndex = NumBitLevels; bitIndex != 0; ) {
+            bitIndex--;
+            int bit = (symbol >>> bitIndex) & 1;
+            rangeEncoder.encode(Models, m, bit);
+            m = (m << 1) | bit;
+        }
+    }
+
+    public void reverseEncode(Encoder rangeEncoder, int symbol) throws IOException {
+        int m = 1;
+        for (int i = 0; i < NumBitLevels; i++) {
+            int bit = symbol & 1;
+            rangeEncoder.encode(Models, m, bit);
+            m = (m << 1) | bit;
+            symbol >>= 1;
+        }
+    }
+
+    public int getPrice(int symbol) {
+        int price = 0;
+        int m = 1;
+        for (int bitIndex = NumBitLevels; bitIndex != 0; ) {
+            bitIndex--;
+            int bit = (symbol >>> bitIndex) & 1;
+            price += Encoder.getPrice(Models[m], bit);
+            m = (m << 1) + bit;
+        }
+        return price;
+    }
+
+    public int reverseGetPrice(int symbol) {
+        int price = 0;
+        int m = 1;
+        for (int i = NumBitLevels; i != 0; i--) {
+            int bit = symbol & 1;
+            symbol >>>= 1;
+            price += Encoder.getPrice(Models[m], bit);
+            m = (m << 1) | bit;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 53:58, 66:71.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.